### ISSUE TITLE
[Agent] Agent seems unable to upgrade from 7.13.4 to 7.14 - error in logs ' Cannot create a file when that file already exists'

### ISSUE THREAD

**Issue by EricDavisX** [elastic/beats#27342]
User tried to update elastic-agent on a Windows 2019 server, but it is failing.  It will go to "updating", then it will go to "unhealthy" for several many minutes, then will finally revert to healthy, but it's still at 7.13.4.  Error message indicates failures circa the download of the new agent or processing around that time.

User reports this from cloud.elastic.co and 7.14 'GA' build (not using a snapshot or recent BC)

First error log snippet:
```
11:11
10:52:58.263
elastic_agent
[elastic_agent][info] 2021-08-12T14:52:58Z - message: Application: [ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to UPDATING: Update to version '7.14.0' started - type: 'STATE' - sub_type: 'UPDATING'
10:54:59.404
elastic_agent
[elastic_agent][error] 2021-08-12T14:54:59Z - message: Application: [ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to FAILED: failed upgrade of agent binary: 2 errors occurred:
	* package 'C:\Program Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip' not found: open C:\Program Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip: The system cannot find the file specified.
	* fetching package failed: context deadline exceeded (Client.Timeout or context cancellation while reading body)
 - type: 'ERROR' - sub_type: 'FAILED'
10:54:59.404
elastic_agent
[elastic_agent][error] failed to dispatch actions, error: failed upgrade of agent binary: 2 errors occurred:
	* package 'C:\Program Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip' not found: open C:\Program Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip: The system cannot find the file specified.
	* fetching package failed: context deadline exceeded (Client.Timeout or context cancellation while reading body)
10:54:59.404
elastic_agent
[elastic_agent][error] Elastic Agent status changed to: 'error'

```
Second error log snippet:
```
elastic_agent
[elastic_agent][info] 2021-08-12T15:13:24Z - message: Application: [ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to UPDATING: Update to version '7.14.0' started - type: 'STATE' - sub_type: 'UPDATING'
11:13:37.593
elastic_agent
[elastic_agent][error] 2021-08-12T15:13:37Z - message: Application: [ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to FAILED: failed to update agent symlink: symlink C:\Program Files\Elastic\Agent\data\elastic-agent-e127fc\elastic-agent.exe C:\Program Files\Elastic\Agent\elastic-agent.exe.prev: Cannot create a file when that file already exists. - type: 'ERROR' - sub_type: 'FAILED'
11:13:37.593
elastic_agent
[elastic_agent][error] failed to dispatch actions, error: failed to update agent symlink: symlink C:\Program Files\Elastic\Agent\data\elastic-agent-e127fc\elastic-agent.exe C:\Program Files\Elastic\Agent\elastic-agent.exe.prev: Cannot create a file when that file already exists.
11:13:37.593
elastic_agent
[elastic_agent][error] Elastic Agent status changed to: 'error'
11:18:09.057
elastic_agent
[elastic_agent][info] Elastic Agent status changed to: 'online'
```

**Comment by patrickt333** [comment]
Please note the different error messages.  First, it failed to download the zip file (or couldn't find it, bad path, maybe?)

Second, it found the zip file, but said it couldn't overwrite the executable.

**Comment by blakerouse** [comment]
First I want to say I am very happy to see that it re-checks in with the old version and remains healthy after that point. I understand that it's not the correct version, but that is a good sign that it did not just leave the agent in a bad state and it never comes back online.

So the first error that is reported is because it timed out downloading the binary from the Elastic, this is probably due to the speed of the connection and it cannot download it in time. Seems that it tried again and did download in the needed amount of time, leading to the second error.

The second error looks like an issue that was fixed in the case of a bad upgrade the retry of the upgrade would fail. I believe this was originally reported in issue #26513. It seems that it was fixed and should be in 7.14.

@michalpristas Can you provide some insight on this? Seems that this should be fixed and the error message is identical.

**Comment by michalpristas** [comment]
@blakerouse  is correct in both cases. the workaround for first one is to increase timeout and restart agent. timeout is increased using `agent.download.timeout` by default it's 5 minutes if i recall correctly
increasing it to 10 minutes or more could be enough

workaround for second one is to remove `.prev` file manually before proceeding with upgrade

**Comment by patrickt333** [comment]
So, I can't find that file.  Well, I sort of do, but it's as a symlink to
elastic-agent.exe and windows can't find the symlink file. So:

Do a search in Explorer, and you'll find this:

[image: image.png]

When you click to open the file location, Windows can't find it.  If I look
at the properties, it says it's pointing to C:\Program
Files\Elastic\Agent\data\elastic-agent-054e22 but that directory doesn't
exist.  FWIW, I have two elastic-agent.exe in the Agent root, both pointing
to this non-existent directory.

I see the same in a cmd prompt.
C:\Program Files\Elastic>dir /s | findstr prev
07/28/2021  09:10 PM    <SYMLINK>      elastic-agent.exe.prev [C:\Program
Files\Elastic\Agent\data\elastic-agent-054e22\elastic-agent.exe]

And when I tried the update again, I get yet another error in the log about
hash values but, again, it does come back as a healthy 7.13.4 agent after 5
minutes:

09:42:24.664
elastic_agent
[elastic_agent][info] 2021-08-13T13:42:24Z - message: Application:
[ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to UPDATING: Update
to version '7.14.0' started - type: 'STATE' - sub_type: 'UPDATING'
09:42:25.867
elastic_agent
[elastic_agent][error] 2021-08-13T13:42:25Z - message: Application:
[ed1caab6-be5d-4d4c-87de-316a6a8783a2]: State changed to FAILED: failed
verification of agent binary: 2 errors occurred: * hash for
'elastic-agent-7.14.0-windows-x86_64.zip' not found in 'C:\Program
Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip.sha512'
* hash for 'elastic-agent-7.14.0-windows-x86_64.zip' not found - type:
'ERROR' - sub_type: 'FAILED'
09:42:25.867
elastic_agent
[elastic_agent][error] failed to dispatch actions, error: failed
verification of agent binary: 2 errors occurred: * hash for
'elastic-agent-7.14.0-windows-x86_64.zip' not found in 'C:\Program
Files\Elastic\Agent\data\elastic-agent-1907c2\downloads\elastic-agent-7.14.0-windows-x86_64.zip.sha512'
* hash for 'elastic-agent-7.14.0-windows-x86_64.zip' not found
09:42:25.867
elastic_agent
[elastic_agent][error] Elastic Agent status changed to: 'error'
09:47:07.996
elastic_agent
[elastic_agent][info] Elastic Agent status changed to: 'online'

**Comment by EricDavisX** [comment]
@michalpristas @michel-laterman - maybe Michel, would you mind self-assigning and taking care to make sure about this?  I *maybe got confused because I hear it fixes were merged to 7.14, which to me implies the 'GA' 7.14.0 but I think it was meant that they were only merged to 7.14.1 which is not shipped yet so it can't be tested except with upgrade using a snapshot Kibana of 7.14.1 branch...

**Comment by blakerouse** [comment]
@EricDavisX this change was landed in 7.14 GA, it does not need a 7.14.1 to fix it.

We could by default increase the timeout to try to prevent the timeout of download, but 5 minutes is already a long time. But if the internet is slow because its in another country or something, it might not be enough.

**Comment by ruflin** [comment]
Is there a way that a user could overwrite this timeout with a higher value?

**Comment by blakerouse** [comment]
They can locally on the host, don't think its an option in Fleet UI.

`agent.download.timeout` is the setting.

**Comment by patrickt333** [comment]
Do we have a workaround I can use?  Or is just uninstall and reinstall at
this point?

**Comment by blakerouse** [comment]
@patrickt333 Could you try to set the `agent.download.timeout` to `10m` in the `elastic-agent.yml` and restart the Elastic Agent, then try to perform the upgrade?

**Comment by patrickt333** [comment]
Will try this when im back from vacay next week.

**Comment by EricDavisX** [comment]
I pinged Patrick off line.  And we'll try to re-test this on the manual test team side, too.  @amolnater-qasource @dikshachauhan-qasource this may have some context for the 7.13.4 -> 7.14 -> 7.15 upgrade I asked us to re-confirm.

**Comment by blakerouse** [comment]
I have tested upgrading of:

Windows 10 - 7.13.4 -> 7.14.0 (Success)
Windows Server 2019 Datacenter (GCP) - 7.13.4 -> 7.14.0 (Success)

I have not been able to reproduce the issue. I think it is related to the initial failed download, that places it in a bad state. Once in that bad state for some reason Elastic Agent cannot recover from it. I have not been able to isolate why at the moment.

**Comment by blakerouse** [comment]
Cleaning up the temp directory seems to have fixed the issue and the upgrade was able to get past its current point. I believe that issue to be fixed in 7.15.0 with the change I made here:

https://github.com/elastic/beats/pull/27396/files#diff-70d8a823ae95f40da4d34d3bb54f34c9f6a11b88a914c0abd84ca7630e1a2aafR48

**Comment by EricDavisX** [comment]
So we can re-test in 7.15, sounds ok to me.  @patrickt333 the latest snapshots should work to test if you were inclined to do so.

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have revalidated Agent upgrade on 7.15.0 Snapshot with 7.14.1 released agent and 7.13.4 released agent.
We have installed both version Windows agents with Policies having System and Endpoint Security integrations.
- Both the Agents `v7.14.1` and `v7.13.4` upgrades to 7.15.0 Snapshot successfully.
- Data for System and Endpoint Security generates at regular interval of time after Agent upgrade.

**Build details:**
Build: 43884
Commit 68ba670440494f5c58588b0473b75c6607276c45

Please let us know if anything else is required from our end.
Thanks

**Comment by EricDavisX** [comment]
This is great @amolnater-qasource thanks.  Patrick, if you want to test it and around this that would be great - but for now I'm closing it out.
